File: solutions/python/food-chain/1/food_chain.py
# def recite(start_verse, end_verse):
#     pass


verse_fly = ["I know an old lady who swallowed a fly.", "I don't know why she swallowed the fly. Perhaps she'll die."]

def line_break():
    print("\n")


def fly():
    for line in verse_fly:
        print(line)
    line_break()




verse_spider = ["I know an old lady who swallowed a spider.", "It wriggled and jiggled and tickled inside her.", "She swallowed the spider to catch the fly.", "I don't know why she swallowed the fly. Perhaps she'll die."]


def spider():
    for line in verse_spider:
        print(line)
    line_break()




verse_bird = ["I know an old lady who swallowed a bird.", "How absurd to swallow a bird!", "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.", "She swallowed the spider to catch the fly.", "I don't know why she swallowed the fly. Perhaps she'll die."]

def bird():
    for line in verse_bird:
        print(line)
    line_break()


verse_cat = ["I know an old lady who swallowed a cat.", 
"Imagine that, to swallow a cat!", 
"She swallowed the cat to catch the bird.",
"She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.",
"She swallowed the spider to catch the fly.",
"I don't know why she swallowed the fly. Perhaps she'll die."]

def cat():
    for line in verse_cat:
        print(line)
    line_break()

verse_dog = ["I know an old lady who swallowed a dog.", 
"What a hog, to swallow a dog!",
"She swallowed the dog to catch the cat.",
"She swallowed the cat to catch the bird.", 
"She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.",
"She swallowed the spider to catch the fly.",
"I don't know why she swallowed the fly. Perhaps she'll die."]

def dog():
    for line in verse_dog:
        print(line)
    line_break()



verse_goat = ["I know an old lady who swallowed a goat.", "Just opened her throat and swallowed a goat!", "She swallowed the goat to catch the dog.", "She swallowed the dog to catch the cat.", "She swallowed the cat to catch the bird.", "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.", "She swallowed the spider to catch the fly.", "I don't know why she swallowed the fly. Perhaps she'll die."]


def goat():
    for line in verse_goat:
        print(line)
    line_break()


verse_cow = ["I know an old lady who swallowed a cow.", "I don't know how she swollowed a cow!", "She swallowed the cow to catch the goat.", "She swallowed the goat to catch the dog.", "She swallowed the dog to catch the cat.", "She swallowed the cat to catch the bird", "She swallowed the bird to catch the spider that wriggled and jiggled and tickled in side her.", "She swallowed the spider to catch the fly", "I don't know why she swallowed the fly. Perhaps she'll die."]

def cow():
    for line in verse_cow:
        print(line)
    line_break()


verse_horse = ["I know an old lady who swallowed a horse.", "She's dead, of course!"]

def horse():
    for line in verse_horse:
        print(line)
    line_break()


def song():
    fly()
    spider()
    bird()
    cat()
    dog()
    goat()
    cow()
    horse()

song()

File: solutions/ruby/food-chain/1/food_chain.rb
module Lyrics

  VERSE = <<~verse
    %<food>s
    %<reason>s
    %<consequence>s
  verse

  FOOD = <<~critter
    I know an old lady who swallowed a %<critter>s.
  critter

  THIS_TO_CATCH_THAT = <<~this_to_catch_that
    She swallowed the %<critter>s to catch the %<previous>s.
  this_to_catch_that

  DEATH_BY_FLY = <<~consequence_fly
    I don't know why she swallowed the fly. Perhaps she'll die.

  consequence_fly

  DEATH_BY_HORSE = <<~consequence_horse
    She's dead, of course!
  consequence_horse

end

class FoodChain
  include Lyrics

  @@this_to_catch_that = +''

  CRITTER = {
    fly: "I don't know this_to_catch_that she swallowed the fly. Perhaps she'll die.",
    spider: "It wriggled and jiggled and tickled inside her.\n",
    bird: "How absurd to swallow a bird!\n",
    cat: "Imagine that, to swallow a cat!\n",
    dog: "What a hog, to swallow a dog!\n",
    goat: "Just opened her throat and swallowed a goat!\n",
    cow: "I don't know how she swallowed a cow!\n",
    horse: "She's dead, of course!\n"
  }

  CRITTERS = CRITTER.keys


  def self.song
    CRITTERS.map.with_index do |critter, i|
      previous = CRITTERS[i - 1]
      new(critter: critter, previous: previous).verse
    end.join
  end

  private

  attr_writer :critter,
              :previous

  attr_accessor :current,
                :spider

  def initialize(critter:, previous: nil)
    self.critter = critter
    self.previous = previous
    self.spider = 'spider that wriggled and jiggled and tickled inside her'
    self.current = +''
  end

  def this_to_catch_that
    if critter == :bird
      self.previous = spider
    else
    end
    @@this_to_catch_that.prepend(THIS_TO_CATCH_THAT % {critter:, previous:})
  end

  public

  attr_reader :critter,
              :previous

  attr_accessor = :current

  def verse
    current << FOOD % {critter:}

    case critter
    when :fly
      current << DEATH_BY_FLY
    when :spider
      this_to_catch_that
      current << CRITTER[critter]
      current << @@this_to_catch_that
      current << DEATH_BY_FLY
    when :horse
      current << DEATH_BY_HORSE
    else
      this_to_catch_that
      current << CRITTER[critter]
      current << @@this_to_catch_that
      current << DEATH_BY_FLY
    end
  end

end

if $PROGRAM_NAME == __FILE__
  puts FoodChain.song

end
